### ISSUE TITLE
Dynamic Role Model partly allows users to create users with more rights

### ISSUE THREAD

**Issue by StepBee** [apache/cloudstack#5781]
##### ISSUE TYPE

 * Bug Report

##### COMPONENT NAME

~~~
API
~~~

##### CLOUDSTACK VERSION

~~~
4.16.0
~~~

##### CONFIGURATION

N/A

##### OS / ENVIRONMENT

N/A

##### SUMMARY

A new role, for example "Domain Admin Restricted", is created by a Root Admin Account based on existing role "Domain Admin"
Some functionality is restricted of the role "Domain Admin Restricted" with setting "Deny" for some APIs, for example createServiceOffering.
A new Account/User, DAJonDoeRestricted, is created based on the role "Domain Admin Restricted".
The user "DAJonDoeRestricted" is not allowed to execute createServiceOffering.
But the user DAJonDoeRestricted is allowed to create a new Account based on the default role "Domain Admin".
For this user, the deny rules of role Domain Admin Restricted are not inherited and do not apply.
The newly created user has higher rights than the restricted user.

##### STEPS TO REPRODUCE

Create Role "Domain Admin Restricted" based on "Domain Admin"
Restrict functionality by choosing Deny for "createServiceOffering"
Create an Account "DAJonDoeRestricted" based on "Domain Admin Restricted"
Login as account "DAJonDoeRestricted"
Create a new account "DAJaneDoeNotrestricted" based on Role "Domain Admin"
Logout user "DAJonDoeRestricted"
Login user "DAJaneDoeNotrestricted"

User "DAJonDoeRestricted" is not allowed to issue createServiceOffering.
User "DAJaneDoeNotrestricted" is allowed to issue createServiceOffering.

##### EXPECTED RESULTS

Accounts should not be able to create accounts with rights for more functionality than they are allowed themselves.
Restrictions should be inherited or the ability to create accounts based on Roles with more functionality rights should not be given.

##### ACTUAL RESULTS

Accounts based restricted roles, which were based on Domain Admin role, are able to create Accounts with original/unrestricted Domain Admin functionality rights.

**Comment by DaanHoogland** [comment]
@StepBee, would this be a reasonable implementation for you:
```
get the current user's role
get the requested role
if there is any right in requested role that the current user doesn't have; refuse
```
I suspect there is other implementation possibilities, so let me know if you see some more bears on the road ;)

**Comment by StepBee** [comment]
@DaanHoogland thanks for taking care of this issue and thinking about an easy solution.

Indeed, refusing to create a user based on a role with more rights than the current user would be an easy and, from my perspective, satisfying solution.
Maybe as a bonus there is a chance of not even displaying such roles on the webui when creating a new user?

**Comment by DaanHoogland** [comment]
@StepBee , I think that the algorithm for checking such escalation happening is not trivial; There is deny and allow to check for each entry, but also wildcards are allowed. I think for listing roles this would/may be a bit much given we have 600+ API by now.
we could just iterate over all API and do a check for the caller to se if they are allowed, but if the list is more than the default eight roles, ....
I agree with the use-case though ;)

**Comment by rohityadavcloud** [comment]
Thanks for reporting this @StepBee - generally, when there's a domain admin account we assume the typical use-case is in an enterprise where the domain admin is privileged in some way. But we should prevent and ensure a role can't create account with roles more than their own role priviledge.

**Comment by StepBee** [comment]
@rohityadavcloud i agree.
In my case we have a multi-tenant cloud platform and customers with domain admin privilege for their customer-domain to be able to create users/accounts, but they should not be able to create for example Service Offerings, as our customers are not aware of the required host / storage tags.
Thats why i tried to configure a domain admin for customers, with carefully restricted rights, where this fix will help a lot.

**Comment by borisstoyanov** [comment]
I've tested the fix and it seems to restrict the 'restricted domain admin' to create a normal domain admin, @StepBee can you check if this solves it for you as well.

**Comment by sureshanaparti** [comment]
> I've tested the fix and it seems to restrict the 'restricted domain admin' to create a normal domain admin, @StepBee can you check if this solves it for you as well.

ping @StepBee can you check / confirm the fix.

**Comment by StepBee** [comment]
unfortunately, i am currently in lack of a properly working test environment i could use for testing.
To not cause any delay i would trust @borisstoyanov result

**Comment by sureshanaparti** [comment]
Fixed in #5879. @StepBee If you still notice this issue after the fix, please reopen (with the details). Thanks.

**Comment by StepBee** [comment]
Many thanks for your work on this fix, much appreciated.
Yep, will test in my test envrionment once functional again and reopen in case i see issues.
